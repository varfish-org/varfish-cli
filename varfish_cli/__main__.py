"""Main entry point for VarFish CLI."""

import argparse
import logging
import os
import sys

import logzero
from logzero import logger
import toml

from varfish_cli import __version__
from varfish_cli.case import run as run_case
from varfish_cli.case import setup_argparse as setup_argparse_case
from varfish_cli.common import CommonConfig, run_nocmd
from varfish_cli.varannos import run as run_varannos
from varfish_cli.varannos import setup_argparse as setup_argparse_varannos

#: Paths to search the global configuration in.
GLOBAL_CONFIG_PATHS = ("~/.varfishrc.toml",)


def setup_argparse_only():  # pragma: nocover
    """Wrapper for ``setup_argparse()`` that only returns the parser.

    Only used in sphinx documentation via ``sphinx-argparse``.
    """
    return setup_argparse()[0]


def setup_argparse():
    """Create argument parser."""
    # Construct argument parser and set global options.
    parser = argparse.ArgumentParser(prog="varfish-cli")
    parser.add_argument("--verbose", action="store_true", default=False, help="Increase verbosity.")
    parser.add_argument("--version", action="version", version="%%(prog)s %s" % __version__)

    group = parser.add_argument_group("Basic Configuration")
    group.add_argument(
        "--no-verify-ssl",
        dest="verify_ssl",
        default=True,
        action="store_false",
        help="Disable HTTPS SSL verification",
    )
    group.add_argument(
        "--config",
        default=os.environ.get("VARFISH_CONFIG_PATH", None),
        help="Path to configuration file.",
    )
    group.add_argument(
        "--varfish-server-url",
        default=os.environ.get("VARFISH_SERVER_URL", None),
        help="VarFish server URL key to use, defaults to env VARFISH_SERVER_URL.",
    )
    group.add_argument(
        "--varfish-api-token",
        default=os.environ.get("VARFISH_API_TOKEN", None),
        help="VarFish API token to use, defaults to env VARFISH_API_TOKEN.",
    )

    # Add sub parsers for each argument.
    subparsers = parser.add_subparsers(dest="cmd")

    setup_argparse_case(subparsers.add_parser("case", help="Work with cases."))
    setup_argparse_varannos(subparsers.add_parser("varannos", help="Work with varannos module."))

    return parser, subparsers


def main(argv=None):
    """Main entry point before parsing command line arguments."""
    # Setup command line parser.
    parser, subparsers = setup_argparse()

    # Actually parse command line arguments.
    args = parser.parse_args(argv)

    # Setup logging incl. verbosity.
    if args.verbose:  # pragma: no cover
        level = logging.DEBUG
    else:
        # Remove module name and line number if not running in debug mode.s
        formatter = logzero.LogFormatter(
            fmt="%(color)s[%(levelname)1.1s %(asctime)s]%(end_color)s %(message)s"
        )
        logzero.formatter(formatter)
        level = logging.INFO
    logzero.loglevel(level=level)

    # Load configuration, if any.
    if args.config:
        config_paths = (args.config,)
    else:
        config_paths = GLOBAL_CONFIG_PATHS
    for config_path in config_paths:
        config_path = os.path.expanduser(os.path.expandvars(config_path))
        if os.path.exists(config_path):
            with open(config_path, "rt") as tomlf:
                toml_config = toml.load(tomlf)
            break
    else:
        toml_config = None
        logger.info("Could not find any of the global configuration files %s.", config_paths)

    # Merge configuration from command line/environment args and configuration file.
    config = CommonConfig.create(args, toml_config)

    # Handle the actual command line.
    cmds = {None: run_nocmd, "case": run_case, "varannos": run_varannos}

    res = cmds[args.cmd](
        config, toml_config, args, parser, subparsers.choices[args.cmd] if args.cmd else None
    )
    if not res:
        logger.info("All done. Have a nice day!")
    else:  # pragma: nocover
        logger.error("Something did not work out correctly.")
    return res


if __name__ == "__main__":  # pragma: no cover
    sys.exit(main(sys.argv))
