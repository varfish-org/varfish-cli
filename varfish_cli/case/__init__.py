"""Implementation of varfish-cli subcommand 'case'."""

import argparse

from varfish_cli.case.config import CaseConfig, OutputFormat
from varfish_cli.case.create_case_import import (
    setup_argparse as setup_argparse_create_import,
)
from varfish_cli.case.list_case import setup_argparse as setup_argparse_list
from varfish_cli.case.list_case_import import (
    setup_argparse as setup_argparse_list_import,
)
from varfish_cli.case.small_var_query_create import (
    setup_argparse as setup_argparse_small_var_query_create,
)
from varfish_cli.case.small_var_query_fetch_results import (
    setup_argparse as setup_argparse_small_var_query_fetch_results,
)
from varfish_cli.case.small_var_query_list import (
    setup_argparse as setup_argparse_small_var_query_list,
)
from varfish_cli.case.small_var_query_retrieve import (
    setup_argparse as setup_argparse_small_var_query_retrieve,
)
from varfish_cli.case.small_var_query_settings_shortcut import (
    setup_argparse as setup_argparse_small_var_query_settings_shortcut,
)
from varfish_cli.case.small_var_query_status import (
    setup_argparse as setup_argparse_small_var_query_status,
)
from varfish_cli.case.small_var_query_update import (
    setup_argparse as setup_argparse_small_var_query_update,
)
from varfish_cli.common import run_nocmd


def setup_argparse(parser: argparse.ArgumentParser) -> None:
    """Main entry point for subcommand."""
    of_choices = [o.value for o in OutputFormat]
    of_default = OutputFormat.TABLE.value
    parser.add_argument(
        "--output-format",
        help=f"Output format, one of {of_choices}, default: {of_default}",
        choices=of_choices,
        default=of_default,
    )
    parser.add_argument("--output-fields", help="Output fields if non-default ones")
    parser.add_argument(
        "--output-delimiter", help="Separator for CSV output, default: ','", default=","
    )
    parser.add_argument(
        "--output-file", help="Path to file to write to, defaults to stdout", default="-"
    )

    subparsers = parser.add_subparsers(dest="case_cmd")
    setup_argparse_list(subparsers.add_parser("list", help="List cases."))
    setup_argparse_list_import(
        subparsers.add_parser("list-import-info", help="List case import infos.")
    )
    setup_argparse_create_import(
        subparsers.add_parser("create-import-info", help="Create case import infos.")
    )
    setup_argparse_small_var_query_create(
        subparsers.add_parser("small-var-query-create", help="Create small var query")
    )
    setup_argparse_small_var_query_fetch_results(
        subparsers.add_parser("small-var-query-fetch-results", help="Fetch small var query results")
    )
    setup_argparse_small_var_query_list(
        subparsers.add_parser("small-var-query-list", help="List small var queries")
    )
    setup_argparse_small_var_query_retrieve(
        subparsers.add_parser("small-var-query-retrieve", help="Retrieve small variant query")
    )
    setup_argparse_small_var_query_status(
        subparsers.add_parser("small-var-query-status", help="Get small variant query status")
    )
    setup_argparse_small_var_query_update(
        subparsers.add_parser("small-var-query-update", help="Update small variant query")
    )
    setup_argparse_small_var_query_settings_shortcut(
        subparsers.add_parser(
            "small-var-query-shortcut", help="Generate query parameters from shortcut"
        )
    )


def run(config, toml_config, args, parser, subparser):
    """Main entry point for case command."""
    if not args.case_cmd:  # pragma: nocover
        return run_nocmd(config, args, parser, subparser)
    else:
        config = CaseConfig.create(args, config, toml_config)
        return args.case_cmd(config, toml_config, args, parser, subparser)
