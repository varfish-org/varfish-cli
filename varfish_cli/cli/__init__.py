"""Code for main entry point."""

import logging
import typing

import logzero
import typer

from varfish_cli import __version__
from varfish_cli.cli import cases, importer, projects, tools, varannos
from varfish_cli.config import CommonOptions, load_config
from varfish_cli.exceptions import InvalidConfiguration

#: Paths to search the global configuration in.
DEFAULT_PATH_VARFISHRC = "~/.varfishrc.toml"


def version_callback(value: bool):
    """Callback when called with 'version' or '--version'"""
    if value:
        print(f"varfish-cli {__version__}")
        raise typer.Exit()


#: Main CLI ``Typer`` object.
app = typer.Typer(no_args_is_help=True, pretty_exceptions_enable=False)

# Register all sub commands.
app.add_typer(varannos.app, name="varannos", help="Subcommands for 'varannos' API")
app.add_typer(projects.app, name="projects", help="Subcommands for 'project' API")
app.add_typer(cases.app, name="cases", help="Subcommands for 'cases' API")
app.add_typer(importer.app, name="importer", help="Subcommands for 'importer' API")
app.add_typer(tools.app, name="tools", help="Subcommands for 'tools' API")


@app.command("version")
def main_version():
    """Print version as "varfish-cli $version"."""
    version_callback(True)


@app.callback()
def main(
    ctx: typer.Context,
    version: typing.Annotated[
        typing.Optional[bool], typer.Option("--version", callback=version_callback)
    ] = None,
    verbose: typing.Annotated[
        bool, typer.Option("--verbose", "-v", help="Enable verbose output")
    ] = False,
    verify_ssl: typing.Annotated[
        bool, typer.Option("--verify-ssl/--no-verify-ssl", help="Disable SSL verification")
    ] = True,
    config_path: typing.Annotated[
        str,
        typer.Option("--config-path", help="Path to configuration file", envvar="VARFISH_RC_PATH"),
    ] = DEFAULT_PATH_VARFISHRC,
    varfish_server_url: typing.Annotated[
        typing.Optional[str],
        typer.Option(
            "--varfish-server-url",
            help=(
                "VarFish server URL key to use, defaults to env VARFISH_SERVER_URL or read "
                "from configfile"
            ),
            envvar="VARFISH_SERVER_URL",
        ),
    ] = None,
    varfish_api_token: typing.Annotated[
        typing.Optional[str],
        typer.Option(
            "--varfish-server-url",
            help=(
                "VarFish API token to use, defaults to env VARFISH_API_TOKEN or read from "
                "configfile"
            ),
            envvar="VARFISH_API_TOKEN",
        ),
    ] = None,
):
    """Callback for main entry point

    This function handles the global configuration from configuration file,
    environment variables, and command line (in increasing priority).
    """
    _ = version
    # Setup logging
    if verbose:  # pragma: no cover
        level = logging.DEBUG
    else:
        # Remove module name and line number if not running in debug mode.s
        formatter = logzero.LogFormatter(
            fmt="%(color)s[%(levelname)1.1s %(asctime)s]%(end_color)s %(message)s"
        )
        logzero.formatter(formatter)
        level = logging.INFO
    logzero.loglevel(level=level)

    # Load configuration file
    toml_varfish_server_url, toml_varfish_api_token = load_config(config_path)
    if toml_varfish_server_url and not varfish_server_url:
        varfish_server_url = toml_varfish_server_url
    if toml_varfish_api_token and not varfish_api_token:
        varfish_api_token = toml_varfish_api_token

    if not varfish_server_url or not varfish_api_token:
        raise InvalidConfiguration("Need to specify server URL and API token")

    # Construct common options
    ctx.obj = CommonOptions(
        verbose=verbose,
        verify_ssl=verify_ssl,
        config_path=config_path,
        varfish_server_url=varfish_server_url,
        varfish_api_token=varfish_api_token,
    )


#: Define ``typer`` object that can be called from ``__main__.py``.
typer_click_object = typer.main.get_command(app)
