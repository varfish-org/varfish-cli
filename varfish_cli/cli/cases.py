"""Implementation of varfish-cli subcommand "cases *"."""

import typing
import uuid

import typer

from varfish_cli import api, common
from varfish_cli.cli.common import ListObjects, RetrieveObject
from varfish_cli.common import OutputFormat

#: Default fields for projects.
DEFAULT_FIELDS_CASE: typing.Dict[OutputFormat, typing.Optional[typing.Tuple[str, ...]]] = {
    OutputFormat.TABLE.value: ("sodar_uuid", "name", "index", "members"),
    OutputFormat.CSV.value: None,
    OutputFormat.JSON.value: None,
}

#: The ``Typer`` instance to use for the ``cases`` sub command.
app = typer.Typer(no_args_is_help=True)


@app.command("case-list")
def cli_case_list(
    ctx: typer.Context,
    project_uuid: typing.Annotated[
        uuid.UUID, typer.Argument(..., help="UUID of project to list cases for")
    ],
    output_file: typing.Annotated[
        str, typer.Option("--output-file", help="Path to file to write to")
    ] = "-",
    output_format: typing.Annotated[
        OutputFormat, typer.Option("--output-format", help="Output format")
    ] = OutputFormat.TABLE.value,
    output_delimiter: typing.Annotated[
        str, typer.Option("--output-delimiter", help="Delimiter for CSV output")
    ] = ",",
    output_fields: typing.Annotated[
        typing.Optional[typing.List[str]], typer.Option("--output-fields", help="Output fields")
    ] = None,
):
    """List all Case entries for the"""
    common_options: common.CommonOptions = ctx.obj

    list_objects = ListObjects(api.Case)
    return list_objects.run(
        common_options=common_options,
        callable=api.case_list,
        output_file=output_file,
        output_format=output_format,
        output_delimiter=output_delimiter,
        output_fields=output_fields,
        parent_uuid=project_uuid,
        default_fields=DEFAULT_FIELDS_CASE,
    )


@app.command("case-retrieve")
def cli_case_retrieve(
    ctx: typer.Context,
    object_uuid: typing.Annotated[
        uuid.UUID, typer.Argument(..., help="UUID of the object to retrieve")
    ],
    output_file: typing.Annotated[
        str, typer.Option("--output-file", help="Path to file to write to")
    ] = "-",
):
    """Retrieve Case by UUID"""
    common_options: common.CommonOptions = ctx.obj

    retrieve_object = RetrieveObject(api.Case)
    return retrieve_object.run(
        common_options=common_options,
        callable=api.case_retrieve,
        key_name="case_uuid",
        object_uuid=object_uuid,
        output_file=output_file,
    )
