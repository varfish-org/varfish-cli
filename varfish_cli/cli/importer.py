"""Implementation of varfish-cli subcommand "importer *"."""

import sys
import typing
import uuid

from logzero import logger
import typer

from varfish_cli import api, common

#: The ``Typer`` instance to use for the ``cases`` sub command.
app = typer.Typer(no_args_is_help=True)


@app.command("caseimportinfo-list")
def run(
    ctx: typer.Context,
    project_uuid: typing.Annotated[
        uuid.UUID, typer.Argument(..., help="UUID of project to list cases for")
    ],
    owner: typing.Annotated[
        typing.Optional[str],
        typer.Option("--owner", help="Optionally, name of owner to filter for"),
    ] = None,
    output_file: typing.Annotated[
        str, typer.Option("--output-file", help="Path to file to write to")
    ] = "-",
):
    # config, toml_config, args, _parser, _subparser, file=sys.stdout):
    """List case import infos."""
    common_options: common.CommonConfig = ctx.obj
    logger.info("Listing CaseImportInfo records")
    if owner:
        logger.info("- filter owner: %s" % owner)
    else:
        logger.info("- filter owner: <any>")

    res = api.case_import_info_list(
        server_url=common_options.varfish_server_url,
        api_token=common_options.varfish_api_token.get_secret_value(),
        project_uuid=project_uuid,
        owner=owner,
        verify_ssl=common_options.verify_ssl,
    )

    def do_print(file: typing.TextIO):
        print("Case Import Info List", file=file)
        print("=====================", file=file)
        print(file=file)
        if not res:
            print("No records found.", file=file)
        for info in res:
            print("- uuid: %s" % repr(str(info.sodar_uuid)), file=file)
            print("  release: %s" % repr(info.release.value), file=file)
            print("  owner: %s" % repr(info.owner), file=file)
            print("  name: %s" % repr(info.name), file=file)
            print("  index: %s" % repr(info.index), file=file)
            if info.pedigree:
                print("  members:", file=file)
                for member in info.pedigree:
                    print(
                        "    - { name: %s, father: %s, mother: %s, sex: %s, affected: %s }"
                        % tuple(
                            [
                                repr(x)
                                for x in (
                                    member.name,
                                    member.father,
                                    member.mother,
                                    member.sex,
                                    member.affected,
                                )
                            ]
                        ),
                        file=file,
                    )
            else:
                print("  members: []", file=file)
            print(file=file)
        file.flush()

    logger.info("Writing output")
    if output_file == "-":
        do_print(sys.stdout)
    else:
        with open(output_file, "wt") as outputf:
            do_print(outputf)

    logger.info("All done. Have a nice day!")
