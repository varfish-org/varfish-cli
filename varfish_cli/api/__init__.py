"""Code for accessing VarFish Server API."""

from varfish_cli.api.case import *  # noqa: F403, F401
from varfish_cli.api.models import *  # noqa: F403, F401
from varfish_cli.api.varannos import *  # noqa: F403, F401
