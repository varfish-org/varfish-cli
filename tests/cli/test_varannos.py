"""Test CLI for varannos API."""

import json
import typing
import uuid

import pytest
from pytest_mock import MockerFixture
from requests_mock.mocker import Mocker as RequestsMocker
from syrupy import SnapshotAssertion
from typer.testing import CliRunner

from tests.conftest import FakeFs
from varfish_cli.cli import app


@pytest.fixture
def varannoset_list_result_empty() -> typing.List[typing.Any]:
    return []


def test_varannoset_list_empty(
    runner: CliRunner,
    fake_fs_configured: FakeFs,
    requests_mock: RequestsMocker,
    fake_conn: typing.Tuple[str, str],
    varannoset_list_result_empty,
    mocker: MockerFixture,
):
    mocker.patch("varfish_cli.config.open", fake_fs_configured.open_, create=True)
    mocker.patch("varfish_cli.config.os", fake_fs_configured.os)

    project_uuid = str(uuid.uuid4())
    host, token = fake_conn
    m = requests_mock.get(
        f"{host}/varannos/api/varannoset/list-create/{project_uuid}",
        json=varannoset_list_result_empty,
        request_headers={"Authorization": f"Token {token}"},
    )
    _ = m
    result = runner.invoke(
        app, ["--verbose", "varannos", "varannoset-list", "--output-format=json", project_uuid]
    )

    mocker.stopall()

    assert result.exit_code == 0, result.output
    assert result.output == "[]\n"


@pytest.fixture
def varannoset_list_result_one_elements() -> typing.List[typing.Any]:
    with open("tests/cli/data/varannos_varannoset-list.len-1.json", "rt") as inputf:
        return json.load(inputf)


def test_varannoset_list_one_element(
    runner: CliRunner,
    fake_fs_configured: FakeFs,
    requests_mock: RequestsMocker,
    fake_conn: typing.Tuple[str, str],
    varannoset_list_result_one_elements,
    snapshot: SnapshotAssertion,
    mocker: MockerFixture,
):
    mocker.patch("varfish_cli.config.open", fake_fs_configured.open_, create=True)
    mocker.patch("varfish_cli.config.os", fake_fs_configured.os)

    project_uuid = str(uuid.uuid4())
    host, token = fake_conn
    m = requests_mock.get(
        f"{host}/varannos/api/varannoset/list-create/{project_uuid}",
        json=varannoset_list_result_one_elements,
        request_headers={"Authorization": f"Token {token}"},
    )
    _ = m
    result = runner.invoke(
        app, ["--verbose", "varannos", "varannoset-list", "--output-format=json", project_uuid]
    )

    mocker.stopall()

    assert result.exit_code == 0, result.output
    assert result.output == snapshot


def test_varannoset_create(
    runner: CliRunner,
    fake_fs_configured: FakeFs,
    requests_mock: RequestsMocker,
    fake_conn: typing.Tuple[str, str],
    varannoset_list_result_one_elements,
    snapshot: SnapshotAssertion,
    mocker: MockerFixture,
):
    mocker.patch("varfish_cli.config.open", fake_fs_configured.open_, create=True)
    mocker.patch("varfish_cli.config.os", fake_fs_configured.os)

    obj_json = varannoset_list_result_one_elements[0]
    host, token = fake_conn
    m = requests_mock.post(
        f"{host}/varannos/api/varannoset/list-create/{obj_json['project']}",
        json=obj_json,
        request_headers={"Authorization": f"Token {token}"},
    )
    _ = m
    result = runner.invoke(
        app,
        ["--verbose", "varannos", "varannoset-create", obj_json["project"], json.dumps(obj_json)],
    )

    mocker.stopall()

    assert result.exit_code == 0, result.output
    assert result.output == snapshot


def test_varannoset_retrieve(
    runner: CliRunner,
    fake_fs_configured: FakeFs,
    requests_mock: RequestsMocker,
    fake_conn: typing.Tuple[str, str],
    varannoset_list_result_one_elements,
    snapshot: SnapshotAssertion,
    mocker: MockerFixture,
):
    mocker.patch("varfish_cli.config.open", fake_fs_configured.open_, create=True)
    mocker.patch("varfish_cli.config.os", fake_fs_configured.os)

    obj_json = varannoset_list_result_one_elements[0]
    host, token = fake_conn
    m = requests_mock.get(
        f"{host}/varannos/api/varannoset/retrieve-update-destroy/{obj_json['sodar_uuid']}",
        json=obj_json,
        request_headers={"Authorization": f"Token {token}"},
    )
    _ = m
    result = runner.invoke(
        app, ["--verbose", "varannos", "varannoset-retrieve", obj_json["sodar_uuid"]]
    )

    mocker.stopall()

    assert result.exit_code == 0, result.output
    assert result.output == snapshot


def test_varannoset_update(
    runner: CliRunner,
    fake_fs_configured: FakeFs,
    requests_mock: RequestsMocker,
    fake_conn: typing.Tuple[str, str],
    varannoset_list_result_one_elements,
    snapshot: SnapshotAssertion,
    mocker: MockerFixture,
):
    mocker.patch("varfish_cli.config.open", fake_fs_configured.open_, create=True)
    mocker.patch("varfish_cli.config.os", fake_fs_configured.os)

    obj_json = varannoset_list_result_one_elements[0]
    host, token = fake_conn
    m = requests_mock.patch(
        f"{host}/varannos/api/varannoset/retrieve-update-destroy/{obj_json['sodar_uuid']}",
        json=obj_json,
        request_headers={"Authorization": f"Token {token}"},
    )
    _ = m
    result = runner.invoke(
        app,
        [
            "--verbose",
            "varannos",
            "varannoset-update",
            obj_json["sodar_uuid"],
            json.dumps(obj_json),
        ],
    )

    mocker.stopall()

    assert result.exit_code == 0, result.output
    assert result.output == snapshot


def test_varannoset_destroy(
    runner: CliRunner,
    fake_fs_configured: FakeFs,
    requests_mock: RequestsMocker,
    fake_conn: typing.Tuple[str, str],
    varannoset_list_result_one_elements,
    mocker: MockerFixture,
):
    mocker.patch("varfish_cli.config.open", fake_fs_configured.open_, create=True)
    mocker.patch("varfish_cli.config.os", fake_fs_configured.os)

    obj_uuid = varannoset_list_result_one_elements[0]["sodar_uuid"]
    host, token = fake_conn
    m = requests_mock.delete(
        f"{host}/varannos/api/varannoset/retrieve-update-destroy/{obj_uuid}",
        request_headers={"Authorization": f"Token {token}"},
    )
    _ = m
    result = runner.invoke(
        app,
        ["--verbose", "varannos", "varannoset-delete", obj_uuid],
    )

    mocker.stopall()

    assert result.exit_code == 0, result.output


@pytest.fixture
def varannosetentry_list_result_empty() -> typing.List[typing.Any]:
    return []


def test_varannosetentry_list_empty(
    runner: CliRunner,
    fake_fs_configured: FakeFs,
    requests_mock: RequestsMocker,
    fake_conn: typing.Tuple[str, str],
    varannosetentry_list_result_empty,
    mocker: MockerFixture,
):
    mocker.patch("varfish_cli.config.open", fake_fs_configured.open_, create=True)
    mocker.patch("varfish_cli.config.os", fake_fs_configured.os)

    project_uuid = str(uuid.uuid4())
    host, token = fake_conn
    m = requests_mock.get(
        f"{host}/varannos/api/varannosetentry/list-create/{project_uuid}",
        json=varannosetentry_list_result_empty,
        request_headers={"Authorization": f"Token {token}"},
    )
    _ = m
    result = runner.invoke(
        app, ["--verbose", "varannos", "varannosetentry-list", "--output-format=json", project_uuid]
    )

    mocker.stopall()

    assert result.exit_code == 0, result.output
    assert result.output == "[]\n"


def test_varannosetentry_create(
    runner: CliRunner,
    fake_fs_configured: FakeFs,
    requests_mock: RequestsMocker,
    fake_conn: typing.Tuple[str, str],
    varannosetentry_list_result_one_elements,
    snapshot: SnapshotAssertion,
    mocker: MockerFixture,
):
    mocker.patch("varfish_cli.config.open", fake_fs_configured.open_, create=True)
    mocker.patch("varfish_cli.config.os", fake_fs_configured.os)

    obj_json = varannosetentry_list_result_one_elements[0]
    host, token = fake_conn
    m = requests_mock.post(
        f"{host}/varannos/api/varannosetentry/list-create/{obj_json['varannoset']}",
        json=obj_json,
        request_headers={"Authorization": f"Token {token}"},
    )
    _ = m
    result = runner.invoke(
        app,
        [
            "--verbose",
            "varannos",
            "varannosetentry-create",
            obj_json["varannoset"],
            json.dumps(obj_json),
        ],
    )

    mocker.stopall()

    assert result.exit_code == 0, result.output
    assert result.output == snapshot


@pytest.fixture
def varannosetentry_list_result_one_elements() -> typing.List[typing.Any]:
    with open("tests/cli/data/varannos_varannosetentry-list.len-1.json", "rt") as inputf:
        return json.load(inputf)


def test_varannosetentry_list_one_element(
    runner: CliRunner,
    fake_fs_configured: FakeFs,
    requests_mock: RequestsMocker,
    fake_conn: typing.Tuple[str, str],
    varannosetentry_list_result_one_elements,
    snapshot: SnapshotAssertion,
    mocker: MockerFixture,
):
    mocker.patch("varfish_cli.config.open", fake_fs_configured.open_, create=True)
    mocker.patch("varfish_cli.config.os", fake_fs_configured.os)

    project_uuid = str(uuid.uuid4())
    host, token = fake_conn
    m = requests_mock.get(
        f"{host}/varannos/api/varannosetentry/list-create/{project_uuid}",
        json=varannosetentry_list_result_one_elements,
        request_headers={"Authorization": f"Token {token}"},
    )
    _ = m
    result = runner.invoke(
        app, ["--verbose", "varannos", "varannosetentry-list", "--output-format=json", project_uuid]
    )

    mocker.stopall()

    assert result.exit_code == 0, result.output
    assert result.output == snapshot


def test_varannosetentry_retrieve(
    runner: CliRunner,
    fake_fs_configured: FakeFs,
    requests_mock: RequestsMocker,
    fake_conn: typing.Tuple[str, str],
    varannosetentry_list_result_one_elements,
    snapshot: SnapshotAssertion,
    mocker: MockerFixture,
):
    mocker.patch("varfish_cli.config.open", fake_fs_configured.open_, create=True)
    mocker.patch("varfish_cli.config.os", fake_fs_configured.os)

    obj_json = varannosetentry_list_result_one_elements[0]
    host, token = fake_conn
    m = requests_mock.get(
        f"{host}/varannos/api/varannosetentry/retrieve-update-destroy/{obj_json['sodar_uuid']}",
        json=obj_json,
        request_headers={"Authorization": f"Token {token}"},
    )
    _ = m
    result = runner.invoke(
        app, ["--verbose", "varannos", "varannosetentry-retrieve", obj_json["sodar_uuid"]]
    )

    mocker.stopall()

    assert result.exit_code == 0, result.output
    assert result.output == snapshot


def test_varannosetentry_update(
    runner: CliRunner,
    fake_fs_configured: FakeFs,
    requests_mock: RequestsMocker,
    fake_conn: typing.Tuple[str, str],
    varannosetentry_list_result_one_elements,
    snapshot: SnapshotAssertion,
    mocker: MockerFixture,
):
    mocker.patch("varfish_cli.config.open", fake_fs_configured.open_, create=True)
    mocker.patch("varfish_cli.config.os", fake_fs_configured.os)

    obj_json = varannosetentry_list_result_one_elements[0]
    host, token = fake_conn
    m = requests_mock.patch(
        f"{host}/varannos/api/varannosetentry/retrieve-update-destroy/{obj_json['sodar_uuid']}",
        json=obj_json,
        request_headers={"Authorization": f"Token {token}"},
    )
    _ = m
    result = runner.invoke(
        app,
        [
            "--verbose",
            "varannos",
            "varannosetentry-update",
            obj_json["sodar_uuid"],
            json.dumps(obj_json),
        ],
    )

    mocker.stopall()

    assert result.exit_code == 0, result.output
    assert result.output == snapshot


def test_varannosetentry_destroy(
    runner: CliRunner,
    fake_fs_configured: FakeFs,
    requests_mock: RequestsMocker,
    fake_conn: typing.Tuple[str, str],
    varannosetentry_list_result_one_elements,
    mocker: MockerFixture,
):
    mocker.patch("varfish_cli.config.open", fake_fs_configured.open_, create=True)
    mocker.patch("varfish_cli.config.os", fake_fs_configured.os)

    obj_uuid = varannosetentry_list_result_one_elements[0]["sodar_uuid"]
    host, token = fake_conn
    m = requests_mock.delete(
        f"{host}/varannos/api/varannosetentry/retrieve-update-destroy/{obj_uuid}",
        request_headers={"Authorization": f"Token {token}"},
    )
    _ = m
    result = runner.invoke(
        app,
        ["--verbose", "varannos", "varannosetentry-delete", obj_uuid],
    )

    mocker.stopall()

    assert result.exit_code == 0, result.output
