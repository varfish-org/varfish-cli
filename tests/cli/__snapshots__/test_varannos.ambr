# serializer version: 1
# name: test_varannoset_create
  '''
  {'sodar_uuid': 'e211747f-2a50-4a65-b192-c96bc2e111fa', 'date_created': '2023-10-31T08:15:15+01:00', 'date_modified': '2023-10-31T08:15:15+01:00', 'project': '062b8838-453f-4cf3-817d-a5ec76546462', 'title': 'my title', 'description': 'None', 'release': 'GRCh37', 'fields': ['pathogenicity', 'notes']}
  {
    "sodar_uuid": "e211747f-2a50-4a65-b192-c96bc2e111fa",
    "date_created": "2023-10-31T08:15:15+01:00",
    "date_modified": "2023-10-31T08:15:15+01:00",
    "project": "062b8838-453f-4cf3-817d-a5ec76546462",
    "title": "my title",
    "description": "None",
    "release": "GRCh37",
    "fields": [
      "pathogenicity",
      "notes"
    ]
  }
  
  '''
# ---
# name: test_varannoset_list_one_element
  '''
  [
    {
      "sodar_uuid": "e211747f2a504a65b192c96bc2e111fa",
      "date_created": "2023-10-31T08:15:15.000000+0100",
      "date_modified": "2023-10-31T08:15:15.000000+0100",
      "project": "062b8838453f4cf3817da5ec76546462",
      "title": "my title",
      "description": "None",
      "release": "GRCh37",
      "fields": [
        "pathogenicity",
        "notes"
      ]
    }
  ]
  
  '''
# ---
# name: test_varannoset_retrieve
  '''
  {
    "sodar_uuid": "e211747f-2a50-4a65-b192-c96bc2e111fa",
    "date_created": "2023-10-31T08:15:15+01:00",
    "date_modified": "2023-10-31T08:15:15+01:00",
    "project": "062b8838-453f-4cf3-817d-a5ec76546462",
    "title": "my title",
    "description": "None",
    "release": "GRCh37",
    "fields": [
      "pathogenicity",
      "notes"
    ]
  }
  
  '''
# ---
# name: test_varannoset_update
  '''
  {
    "sodar_uuid": "e211747f-2a50-4a65-b192-c96bc2e111fa",
    "date_created": "2023-10-31T08:15:15+01:00",
    "date_modified": "2023-10-31T08:15:15+01:00",
    "project": "062b8838-453f-4cf3-817d-a5ec76546462",
    "title": "my title",
    "description": "None",
    "release": "GRCh37",
    "fields": [
      "pathogenicity",
      "notes"
    ]
  }
  
  '''
# ---
# name: test_varannosetentry_create
  '''
  {
    "sodar_uuid": "4a3a96ae-ddd8-4e9f-897f-5fbbadfedad5",
    "date_created": "2023-10-31T09:02:26+01:00",
    "date_modified": "2023-10-31T09:02:26+01:00",
    "varannoset": "0a30d59b-7257-43d1-8964-2a2ac50af7f2",
    "release": "GRCh37",
    "chromosome": "1",
    "start": 100,
    "end": 100,
    "reference": "A",
    "alternative": "T",
    "payload": {
      "pathogenic": "likely-pathogenic",
      "notes": "TEST"
    }
  }
  
  '''
# ---
# name: test_varannosetentry_list_one_element
  '''
  [
    {
      "sodar_uuid": "4a3a96aeddd84e9f897f5fbbadfedad5",
      "date_created": "2023-10-31T09:02:26.000000+0100",
      "date_modified": "2023-10-31T09:02:26.000000+0100",
      "varannoset": "0a30d59b725743d189642a2ac50af7f2",
      "release": "GRCh37",
      "chromosome": "1",
      "start": 100,
      "end": 100,
      "reference": "A",
      "alternative": "T",
      "payload": {
        "pathogenic": "likely-pathogenic",
        "notes": "TEST"
      }
    }
  ]
  
  '''
# ---
# name: test_varannosetentry_retrieve
  '''
  {
    "sodar_uuid": "4a3a96ae-ddd8-4e9f-897f-5fbbadfedad5",
    "date_created": "2023-10-31T09:02:26+01:00",
    "date_modified": "2023-10-31T09:02:26+01:00",
    "varannoset": "0a30d59b-7257-43d1-8964-2a2ac50af7f2",
    "release": "GRCh37",
    "chromosome": "1",
    "start": 100,
    "end": 100,
    "reference": "A",
    "alternative": "T",
    "payload": {
      "pathogenic": "likely-pathogenic",
      "notes": "TEST"
    }
  }
  
  '''
# ---
# name: test_varannosetentry_update
  '''
  {
    "sodar_uuid": "4a3a96ae-ddd8-4e9f-897f-5fbbadfedad5",
    "date_created": "2023-10-31T09:02:26+01:00",
    "date_modified": "2023-10-31T09:02:26+01:00",
    "varannoset": "0a30d59b-7257-43d1-8964-2a2ac50af7f2",
    "release": "GRCh37",
    "chromosome": "1",
    "start": 100,
    "end": 100,
    "reference": "A",
    "alternative": "T",
    "payload": {
      "pathogenic": "likely-pathogenic",
      "notes": "TEST"
    }
  }
  
  '''
# ---
